src/htm/ntypes/ArrayBase.cpp

```diff
@@ -658,13 +658,13 @@ std::string ArrayBase::toJSON() const {
     const SDR &sdr = getSDR();
     json << "[";
     bool first = true;
-    SDR_sparse_t sparse = sdr.getSparse();
-    for (size_t i = 0; i < sparse.size(); i++) {
+    auto dense = sdr.getDense();
+    for (size_t i = 0; i < dense.size(); i++) {
```

Comment:
this is the only thing I'm not too keen on. There could be a performance penalty.

Reply:
Let's talk about this.  I am not keen on it either.  But how will these be used?

In this case we start with an Array object.  Everything contained in an Array object is an array of one type or another but in string format.  All are dense format unless we make the SDR into sparse format.  If we make the SDR type sparse then that forces all users of the REST api to first check the type before they can parse their response into their target destination.   Maybe that is OK.

If the client is C# there is strong typing so the target type will need to be known ahead of time anyway, probably at compile time.  If the client is Python, they would need to know to treat the SDR type response as sparse when loading it into a numpy array to expand to dense.  They will not have SDR objects.

So that is why I made them all dense.

Reply:
this is a good kickoff for an analysis, thank you! 

> Everything contained in an Array object is an array of one type or another but in string format in the JSON. All are dense format 

what kind of data could/typically would be sent in the Array? 
My notion of things is that it's mostly the region's output (being passed as a result or sent to another layer/region). 
What are the other users of (NAPI) Array type? 

> All are dense format unless we make the SDR into sparse format. If we make the SDR type sparse then that forces all users of the REST api to first check the type before they can parse their response 

Ideally, could we switch to the other default (sparse)? That everything in an Array would be sparse data. 

>  If the client is Python,[...] loading it into a numpy array to expand to dense. They will not have SDR objects.

good point, but you could use the SDR in python (?). 

I see your reasoning, if the usecases end up converting through the dense part anyway, it makes sense to make easier the reading part (avoid check if SDR) by defaulting to dense. 

My (not validated) point is not to create a significant disadvantage for NetworkAPI loads. Even currently, "pure" C++ hotgym was 30% faster than python NAPI counterpart. (it would be good to have a performance comparison: pure C++, C++ NAPI, python, python NAPI, optionally even REST ).

Reply:
> what kind of data could/typically would be sent in the Array?

Yes, mostly outputs to inputs.  But not all inputs and outputs are binary data.  An input to an encoder could be an array of any type and cannot be expressed as an SDR.  The 'anomaly' output of the TM is another example of a non-binary array.  If all data were binary I would have disposed of the Array object and just use SDR's everywhere.  The Array object is also returned when passing back parameter data in those cases where the parameter contains an array.

> Ideally, could we switch to the other default (sparse)? That everything in an Array would be sparse data.

That is the point. It is not always binary data and therefore cannot be sparse.

> good point, but you could use the SDR in python (?).

The point of the REST api is that the REST client does NOT have access to the htm.core library.  So SDR objects do not exist. If they were required to load the htm.core library then there would be no reason to use the slower REST interface.  Just call it.  I guess someone could write a pure python SDR object but that is not part of the REST api.

> it would be good to have a performance comparison:

Yes, that would be a nice table. Perhaps including other HTM implementations as well.  But I am not volunteering to build it.  There are plenty of needed fixes to keep me busy.